nvflare/fuel/hci/client/file_transfer.py

```diff
@@ -339,59 +343,113 @@ def download_text_file(self, args, ctx: CommandContext):
             # unzip the file
             api.debug(f"unzipping file {file_path} to {dir_name}")
             os.makedirs(dir_name, exist_ok=True)
             unzip_all_from_file(file_path, dir_name)
+
+            # remove the zip file
+            os.remove(file_path)
         return result
 
     def pull_folder(self, args, ctx: CommandContext):
         cmd_entry = ctx.get_command_entry()
-        if len(args) != 2:
+        if len(args) < 2:
             return {ProtoKey.STATUS: APIStatus.ERROR_SYNTAX, ProtoKey.DETAILS: "usage: {}".format(cmd_entry.usage)}
         folder_name = args[1]
+        destination_name = folder_name
+        if len(args) > 2:
+            destination_name = args[2]
+
         parts = [cmd_entry.full_command_name(), folder_name]
         command = join_args(parts)
         api = ctx.get_api()
         result = api.server_execute(command)
+        if result.get(ProtoKey.STATUS) != APIStatus.SUCCESS:
+            return result
+
         meta = result.get(ProtoKey.META)
         if not meta:
             return result
 
         file_names = meta.get(MetaKey.FILES)
-        ctl_id = meta.get(MetaKey.CONTROL_ID)
-        api.debug(f"received ctl_id {ctl_id}, file names: {file_names}")
+        tx_id = meta.get(MetaKey.TX_ID)
+        api.debug(f"received tx_id {tx_id}, file names: {file_names}")
         if not file_names:
             return result
 
         cmd_name = meta.get(MetaKey.CMD_NAME)
 
-        for file_name in file_names:
-            command = f"{cmd_name} {ctl_id} {file_name}"
+        error = None
+        for i, file_name in enumerate(file_names):
+            parts = [cmd_name, tx_id, folder_name, file_name]
+            if i == len(file_names) - 1:
+                # this is the last file
+                parts.append("end")
+
+            command = join_args(parts)
             reply = api.do_command(command)
             if reply.get(ProtoKey.STATUS) != APIStatus.SUCCESS:
-                return reply
-
-        return {ProtoKey.STATUS: APIStatus.SUCCESS, ProtoKey.DETAILS: "OK"}
+                error = reply
+                break
+
+        if not error:
+            tx_path = self._tx_path(tx_id, folder_name)
+            destination_path = os.path.join(self.download_dir, destination_name)
+            try:
+                os.rename(tx_path, destination_path)
+                location = destination_path
+            except:
+                if destination_name != folder_name:
+                    # try to use alt_destination_path
+                    alt_destination_path = os.path.join(
+                        self.download_dir,
+                        f"{destination_name}__{tx_id}",
+                    )
+                    try:
+                        os.rename(tx_path, alt_destination_path)
+                        location = alt_destination_path
+                    except:
+                        location = tx_path
+                else:
+                    location = tx_path
```

Comment:
If download_job multiple times (sequentially), this will result the job and workspace been downloaded to the "transfer" folder multiple times, each with its own tx_id. This will cause confusion to the users. Most likely the users expect the latest download result to overwrite the previous contents.

Reply:
I disagree. Just like regular download, if the same file is downloaded to your computer's Download folder, each time a new file is created with a different name (usually the name and a suffix).

Reply:
well, probably this is a little subjective. But I found this is very confusing, especially when running the "download_job" command multiple times. 

If we decide to make this change, we need to well document it. We are changing the user experience behavior.

Reply:
Yes this is subjective. On the other hand, it's not easy to support download of the same job at the same time. Both downloads are writing to temp folder, which is not a problem. But at the end of the download, only one of them can be renamed to the proper name, and the other rename will fail.  There is no way to know which download should be kept or removed. That's why I decided to keep both and let the user know exactly where the downloaded content is.

Yes, we'll need to document it.

Reply:
If we want to change this behavior, we should change the "tx_id" to be some number easier to understand, for example, incremental, 1, 2, 3... instead of using the uuid.

Reply:
I guess we could. It would make it look nicer.

Reply:
Fixed - made the folder name ended with simple integers if necessary.